File: Example/DKAudioPlayerExample/ViewController/ViewController.m
//
//  ViewController.m
//  DKAudioPlayerSample
//
//  Created by Denis Kutlubaev on 17.01.14.

#import "ViewController.h"
@import DKAudioPlayer;

@interface ViewController ()

@property (strong, nonatomic) DKAudioPlayer *audioPlayer;
@property (weak, nonatomic) IBOutlet UIView *audioPlayerContainerView;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *audioPlayerContainerViewBottomConstraint;

@end

@implementation ViewController

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];

    if (self) {
        self.tabBarItem.title = @"View";
        self.tabBarItem.image = [UIImage imageNamed:@"42-photos"];
    }

    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self setupAudioPlayer];
}

- (void)setupAudioPlayer {
    NSString *audioFilePath = [[NSBundle mainBundle] pathForResource:@"sample.mp3" ofType:nil];
    if (!audioFilePath) { return; }

    self.audioPlayer = [[DKAudioPlayer alloc] initWithAudioFilePath:audioFilePath];
    [self.audioPlayerContainerView addSubview:self.audioPlayer];
    self.audioPlayer.translatesAutoresizingMaskIntoConstraints = NO;
    [self.audioPlayerContainerView addConstraints:@[
        [self.audioPlayer.leadingAnchor constraintEqualToAnchor:self.audioPlayerContainerView.leadingAnchor],
        [self.audioPlayer.trailingAnchor constraintEqualToAnchor:self.audioPlayerContainerView.trailingAnchor],
        [self.audioPlayer.bottomAnchor constraintEqualToAnchor:self.audioPlayerContainerView.bottomAnchor],
        [self.audioPlayer.topAnchor constraintEqualToAnchor:self.audioPlayerContainerView.topAnchor]
    ]];
}

- (IBAction)showHideButtonTapped:(id)sender {
    // You can show or hide your player as you want
    // This is just an example
    // When you hide it this way, the music doesn't stop
    // If you want to stop a music, use a dismiss method or pause 
    [UIView animateWithDuration:0.4 animations:^{
        if (self.audioPlayerContainerViewBottomConstraint.constant != 0) {
            self.audioPlayerContainerViewBottomConstraint.constant = 0;
        } else {
            CGFloat bottomInset = self.keyWindow.safeAreaInsets.bottom;
            self.audioPlayerContainerViewBottomConstraint.constant = - (self.audioPlayerContainerView.bounds.size.height + bottomInset);
        }
        [self.view setNeedsLayout];
        [self.view layoutIfNeeded];
    }];
}

- (UIWindow *)keyWindow {
    UIWindow *foundWindow = nil;
    NSArray *windows = [[UIApplication sharedApplication] windows];

    for (UIWindow *window in windows) {
        if (window.isKeyWindow) {
            foundWindow = window;
            break;
        }
    }

    return foundWindow;
}

@end

File: Example/DKAudioPlayerExample/ViewController/ViewController.swift
//
//  ViewController.swift
//  DKAudioPlayerExample
//
//  Created by Denis Kutlubaev on 22.11.2022.
//

import DKAudioPlayer
import UIKit

class ViewController: UIViewController {
    private var audioPlayer: DKAudioPlayer?
    @IBOutlet private weak var audioPlayerContainerView: UIView!
    @IBOutlet private weak var audioPlayerContainerViewBottomConstraint: NSLayoutConstraint!

    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)

        tabBarItem.title = "View"
        tabBarItem.image = UIImage(named: "42-photos")
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        setupAudioPlayer()
    }

    func setupAudioPlayer() {
        guard let audioFilePath = Bundle.main.path(forResource: "sample.mp3", ofType: nil) else {
            return
        }

        audioPlayer = DKAudioPlayer(audioFilePath: audioFilePath)
        if let audioPlayer {
            audioPlayerContainerView.addSubview(audioPlayer)
        }
        audioPlayer?.translatesAutoresizingMaskIntoConstraints = false
        audioPlayerContainerView.addConstraints(
            [
                audioPlayer?.leadingAnchor.constraint(equalTo: audioPlayerContainerView.leadingAnchor),
                audioPlayer?.trailingAnchor.constraint(equalTo: audioPlayerContainerView.trailingAnchor),
                audioPlayer?.bottomAnchor.constraint(equalTo: audioPlayerContainerView.bottomAnchor),
                audioPlayer?.topAnchor.constraint(equalTo: audioPlayerContainerView.topAnchor)
            ].compactMap { $0 })
    }

    @IBAction func showHideButtonTapped(_ sender: Any) {
        // You can show or hide your player as you want
        // This is just an example
        // When you hide it this way, the music doesn't stop
        // If you want to stop a music, use a dismiss method or pause
        UIView.animate(withDuration: 0.4, animations: { [self] in
            if audioPlayerContainerViewBottomConstraint.constant != 0 {
                audioPlayerContainerViewBottomConstraint.constant = 0
            } else {
                let bottomInset = keyWindow?.safeAreaInsets.bottom ?? 0.0
                audioPlayerContainerViewBottomConstraint.constant = -(audioPlayerContainerView.bounds.size.height + bottomInset)
            }
            view.setNeedsLayout()
            view.layoutIfNeeded()
        })
    }

    var keyWindow: UIWindow? {
        var foundWindow: UIWindow? = nil
        let windows = UIApplication.shared.windows

        for window in windows {
            if window.isKeyWindow {
                foundWindow = window
                break
            }
        }

        return foundWindow
    }
}
